### ISSUE TITLE
Giant loss for LogSoftmax when Auto-Batching on GPU

### ISSUE THREAD

**Issue by m-hahn** [clab/dynet#747]
Hi,

thank you for the awesome work!

It seems to me that autobatching (on GPU) doesn't work as expected when there is a sum within a logsoftmax.

Here is an MWE (created by simplifying the code where I originally ran into the problem):

```
import dynet as dy
import random

def run():
  b_out = dy.vecInput(46349)
  b_out.set([0] * 46349)
  losses = []
  length = random.randint(5,50)
  for i in range(0, length):
      log_probs = dy.log_softmax(b_out + b_out)
      losses.append(dy.pick(log_probs, random.randint(0,1000)))
  return dy.esum(losses)
           
model = dy.Model()
while True:
  dy.renew_cg()
  losses = []
  for data in range(0,64): 
    losses.append(run())
  total_loss = dy.esum(losses).scalar_value()
  print('loss per sentence: {}'.format(total_loss / 64))
```
When I originally ran into the problem, the sum inside the softmax was an MLP (which is why I need this construct), but it turned out that summing constants was enough to create the problem.

Saving this in session.py and calling
`python session.py --dynet-mem 5000 --dynet-autobatch 1 --dynet-gpu`
I get impossibly huge losses like 1e13 for many (but not all) iterations. Removing either `--dynet-autobatch 1` or `--dynet-gpu`, the loss is about 300, which is expected. When removing `--dynet-mem 5000`, the bad losses turn to NaN.

The problem disappears when changing 
`      log_probs = dy.log_softmax(b_out + b_out)`
to
`      log_probs = dy.log_softmax(b_out)`
b_out is zero, and its dimension is within the usual range in language modeling. Therefore, it is hard for me to imagine how this could be due to numerical problems.

Is there a way of implementing this without running into this difficulty? Or is this a problem inside the autobatching mechanism?

Thank you for any help!

**Comment by neubig** [comment]
Thanks for the report! This looks like it is likely a bug, so we'll try to
figure out the cause of the problem ASAP.

Graham

**Comment by neubig** [comment]
FYI, I was able to reproduce this, so a fix should be coming in short order.

**Comment by m-hahn** [comment]
Do you have an update on the situation? Is it safe to use Auto-Batch + GPU + LogSoftmax, avoiding summation, in the meantime?

Michael

**Comment by neubig** [comment]
FYI: I started looking at this, but the bug is really hard to track down. It only happens when the net is quite large (like the one you are using here), and even then appears for some network structures and not others. I'll continue to debug.

**Comment by neubig** [comment]
@m-hahn Thanks again for tracking down a minimal example for this bug. It should be fixed in #800, so please try the newest version from master.
